keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,956 @@
+graduate through stability levels.
+
+The StorageVersions of an API group will need to be modified to track which
+StorageVersions was used for each Kubernetes version that the compatibility
+version can be set to. This will then be used when the apiserver to write
+resources with the same StorageVersions used by the Kubernetes version the
+compatibility version is set to.
+
+### API Compatibility Versioning
+
+Similar to feature flags, all APIs group-versions declarations will be modified
+to track which Kubernetes version the API group-versions are introduced (or
+removed) at.
+
+GA APIs should match the exact set of APIs enabled in GA for the Kubernetes version
+the compatibility version is set to.
+
+All Alpha and Beta APIs (both off-by-default and on-by-default, if any of those
+still exist) need to behave exactly as they did for the Kubernetes version
+the compatibility version is set to. I.e. `--runtime-config=api/<version>` needs
+to be able to turn on APIs exactly like it did for each Kubernetes version that
+compatibility version can be set to.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A cluster administrator is running Kubernetes 1.30.12 and wishes to perform a
+cautious upgrade to 1.31.5 using the smallest upgrade steps possible, validating
+the health of the cluster between each step.
+
+- For each control plane component, in the [recommended
+  order](https://kubernetes.io/releases/version-skew-policy/):
+  - Upgrades binary to `kubernetes-1.31.5` but sets `--compatibility-version=1.30`
+  - Verifies that the cluster is healthy
+- Next, for each control plane component:
+  - Sets `--compatibility-version=1.31`
+  - Verifies that the cluster is healthy
+
+#### Story 2
+
+A cluster administrator is running Kubernetes 1.30.12 and wishes to perform a cautious
+upgrade to 1.31.5, but after upgrading realizes that a feature the workload depends on
+had been removed and needs to rollback until the workload can be modified to not
+depend on the feature.
+
+- For each control plane component, in the [recommended
+  order](https://kubernetes.io/releases/version-skew-policy/):
+  - sets `--compatibility-version=1.30`
+
+This avoids having to rollback the binary version.  Once the workload is fixed, the
+cluster administrator can remove the `--compatibility-version` to roll the cluster
+forward again.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
```

Comment:
I see an additional risk of unintended and out-of-allowance compatibility skew between binaries.  A kube-apiserver and kube-controller-manager contract is still +/-1 (as far as I see here).  This compatibility level, especially across three versions, makes it more likely for accidental mismatches.

While a hard shutdown of a process is likely worse than the disease, exposing some sort of externally trackable signal for cluster-admins and describing how to use it could significantly mitigate the problem.

Reply:
Adding to risks section.

Possible approaches:

- Clients send versions in headers. Servers use this to detect out-of-allowance skew. Servers then surface this somehow.
- Components register identity leases (apiserver already does this) https://github.com/kubernetes/enhancements/pull/4356 proposes doing it for controller managers.  The components include their version information when they register their leases.  A separate controller inspects all the leases for skew and reports it somehow.

cc @Richabanker who was thinking about adding versions to client headers for UVIP enhancement purposes